Add function to check correctness of `Model`

Lipovr opened the pull request:
# Add function to check correctness of `Model`

A `Model` instance needs to be correct as defined in `piwcsprwmodel.h` 
prior to running complex algorithms such as routing tables generation on it.

This PR adds `bool isComplete(const Model&)` function to check correctness of 
`Model` according to definition in `piwcsprwmodel.h`.

The check is performed in assumption that if all `Node` instances of model are 
_locally correct_ then the model is correct. A node is _locally correct_ if for every section allowing travel __into__ node
there is a valid path to some section allowing travel __from__ node
except for bidirectional ends.

`bool isComplete(const Model&)` iterates through nodes in model and returns `false` upon encountering a non _locally correct_ node, and `true` if all nodes are _locally correct_.

- [x] Code compiles with no warnings using GCC
- [ ] Code compiles with no warnings using clang
- [ ] Code has no clang-tidy errors/warnings
- [x] All tests are up to date
- [x] All tests pass
- [ ] Code is formatted according to `.clang-format`
- [x] Inline documentation is up to date
- [x] Documentation in `include/piwcsprwmodel.h` is up to date
- [x] Doxygen produces no warnings

Lipovr committed:
Added declaration of 'isCorrect'

- Added `isCorrect(const Model &model)`
- Added documentation

Lipovr committed:
Modified documentation

- Modified documentation for `isCorrect(const Model &model)`

Lipovr committed:
Begin implementing `isCorrect`

- Begin implementing `isCorrect(const Model &model)`

Lipovr committed:
Implemented `isCorrect`

- Implemented `isCorrect(const Model &model)`

Lipovr committed:
Finished implementing `isCorrect()`

- Implemented `isCorrect(const Model&)

Lipovr committed:
Added basic tests for `isCorrect`

- Added basic tests for `isCorrect(const Model&)`
- Added test sources to `tests` target

Lipovr committed:
Refined documentation

- Refined documentation for `isComplete(const Model&)

Lipovr committed:
Move local correctness check to separate function

- Move local correctness check to `isLocCorrect(const Model& model, IdRef id)`

OLEGSHA commented on src/algorithms.cpp:
```diff
@@ -22,4 +24,105 @@ bool isComplete(const Model &model) {
+		// Set outward/inward flags
+		for(SlotId start = 0; start < node->sectionCount(); ++start){
+			for(SlotId end = 0; end < node->sectionCount(); ++end){
+				if(
```
I think this could be made more readable if you use boolean flags instead of packing everything inside the if condition

Lipovr replied:
Changed in a6b9f90c3de171f5e65f18c54d3cfd37a8e4bacd

OLEGSHA commented on src/algorithms.cpp:
```diff
@@ -22,4 +24,105 @@ bool isComplete(const Model &model) {
+	const Node* node = model.node(id);
+
+	auto isInward = std::vector<bool>(Node::MAX_SLOTS, false);
+	auto isOutward = std::vector<bool>(Node::MAX_SLOTS, false);
```
Use `std::array<bool, Node::MAX_SLOTS>` instead to improve performance (less dynamic memory allocations/access). `Node::MAX_SLOTS` will never be too big for stack, even if supernodes get implemented

Lipovr replied:
Changed in 97c800b4adaa6188f0b1479151af3b66fcfddb66

OLEGSHA commented on src/algorithms.cpp:
```diff
@@ -22,4 +24,105 @@ bool isComplete(const Model &model) {
     return true;
 }
 
+bool isLocCorrect(const Model& model, IdRef id){
```
1. I'm not sure this is the best name for this function. How about `isLocallyCorrect`?
2. Since this function is not meant to be linked against externally, wrap it in an anonymous namespace

Lipovr replied:
1. Renamed in 537a6a5a98177040cdd047cf470162546280aa7d
2. I think it would be great to test `isLocallyCorrect` separately from `isCorrect`, as this will require example model to only be partially correct. For this purposes created a new private header (537a6a5a98177040cdd047cf470162546280aa7d). As an example added a test for `isLocallyCorrect` (0ccf8cd6827c7677c52f30222378d123d5517309). More tests pending.

OLEGSHA commented on test/CMakeLists.txt:
```diff
@@ -11,6 +11,7 @@ if(GTest_FOUND)
         io_read.cpp
         io_wr.cpp
         completeness.cpp
+		correctness.cpp
```
```suggestion
        correctness.cpp
```
This file isn't processed by formatters, and the indentation was incorrect (changed tabs to spaces)

OLEGSHA reviewed (changes requested):
See comments. More tests are also needed:
- model that is complete but not correct
- unidirectional and forbidden sections
- at least one test per node type
This is more tests than for most of the rest of the library but correctness is a complicated concept, so I think they are warranted

OLEGSHA committed:
Format test/CMakeLists.txt

Lipovr committed:
Changes proposed from https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1577679369

- Rename `isLocCorrect(const Model& model, IdRef id)` to `isLocallyCorrect(const Model& model, IdRef id`
- Move `isLocallyCorrect` and `isCorrect` to correctness.cpp
- Introduce private header for purpose of linking test executable against `isLocallyCorrect`

Lipovr committed:
Remove unnecessary include

Lipovr committed:
Add basic tests for `isLocallyCorrect`

Lipovr committed:
Replaced std::vector with std::array as per https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1577675315

Lipovr commented:
Hello, thanks for Your comments, I will answer them in corresponding threads. As for tests additional ones are pending (hence the "draft" status).

I have some questions regarding "correctness" of a model.

First, is a model with a node that allows travel between two slots ("start" and "end" for example) correct if the section connected to "start" does not allow travelling into that node?

For example, consider a node that is similar in structure to `MOTORIZED` except it allows travel between `STRAIGHT` and `COMMON` in addition to rules of `MOTORIZED`. If then `COMMON` is connected to end slot of unidirectional section and both `STRAIGHT` and `DIVERGING` are connected to start slots of unidirectional sections, and all other parts of model are correct is this model correct? The examined node in this case is locally correct as for every section going into node there is a valid path to section going out of the node.

Secondly, is a model with partially forbidden nodes correct? What I mean is, if for example a `CROSSING` node has forbidden sections in slots `0` and `1` and bidirectional sections in slots `2` and `3` is a model with such node correct (if everything else in it is correct)?

OLEGSHA commented:
Thank you so much for you work on this PR. I'm very sorry for the long delay.

> First, is a model with a node that allows travel between two slots ("start" and "end" for example) correct if the section connected to "start" does not allow travelling into that node?

Yes, it is. Consider a `THRU` node in the middle of a unidirectional line. The sections it is connected to only allow a travel from e.g. 0 to 1. Traveling from 1 to 0 is allowed by the node, but it is forbidden by the sections around it. This is obviously correct.

> Secondly, is a model with partially forbidden nodes correct?

Yes, it is.

Lipovr committed:
Use boolean flags in conditional as proposed in https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1577674418

Lipovr committed:
Add more tests for `isLocallyCorrect`

- Test every node type with all sections forbidden
- Test every node type in at least one correct non forbidden configuration
- Test every node type in at least one incorrect configuration

Lipovr committed:
Refactor & improve readability

OLEGSHA commented on src/correctness.cpp:
```diff
@@ -0,0 +1,120 @@
+				return false;
+			}
+
+			bool isUnidirectional = section->isUnidir();
```
Create the `isUnidirectional` flag a little earlier to make use of it in the check just above. Also, the comment seems to imply the opposite of what's actually happening...
```suggestion
			bool isUnidirectional = section->isUnidir();
			
			// Unidirectional slot
			if(!isUnidirectional){
				return false;
			}
```

Lipovr replied:
Changed in cb58782a014d64ee9e0e9b9ddf00f98707280c2c.

OLEGSHA commented on src/correctness.cpp:
```diff
@@ -0,0 +1,120 @@
+		for(SlotId slot = 0; slot < node->sectionCount(); ++slot){
+			isInward[slot] = false;
+			isOutward[slot] = false;
+		}
```
Why reset these local variables? Is this left over from before `isLocallyCorrect` was extracted?

Lipovr replied:
It is a leftover, adressed in 674f8dc.

OLEGSHA commented on src/correctness.cpp:
```diff
@@ -0,0 +1,120 @@
+
+namespace piwcs::prw {
+
+bool isLocallyCorrect(const Model& model, IdRef id){
```
What are your thoughts on making this function public?

Lipovr replied:
Probably a good idea, will enable, for example, model editing applications to check changed nodes on the go without checking correctness of whole model.

OLEGSHA replied:
I was thinking along the same line. Could you make it public then?

Lipovr replied:
Moved to public header and added documentation in https://github.com/OLEGSHA/PiwcsPrwModel/pull/1/commits/bc046b05dbee9d53399a493692ace0d66b15dfa6.

OLEGSHA commented on src/correctness.cpp:
```diff
@@ -0,0 +1,120 @@
+namespace piwcs::prw {
+
+bool isLocallyCorrect(const Model& model, IdRef id){
+	const Node* node = model.node(id);
```
`node` could be `nullptr` if the ID is bad. Check for that and return `false`, or make a runtime assertion.

Lipovr replied:
Should the function return `false` or raise an exception though? In my opinion exception is better as for a node with invalid id it is impossible to determine if the node is correct. What are your thoughts on that?

OLEGSHA replied:
I was keeping the library exception-free except in IO. In my opinion, the most correct solution would be to add a error return value (e.g. enum IsCorrectResult {YES, NO, ERROR};), but this is far too inconvenient to actually use. We could make the existence of the node a hard requirement (if node does not exist behavior is undefined) and make a runtime assert, but this has a chance of creating a vulnerability somewhere down the line. Returning `false` would hide a potentially sneaky bug in user code.

Weighing the options, I think we should make it return false and document that.

Lipovr replied:
Agreed. Added check for `nullptr` in b451f775411a2fb4df7d5a4ad19f89539d5d241d. Will add documentation for  `isLocallyCorrect` as a separate commit later.

Lipovr replied:
I have documented `isLocallyCorrect` in https://github.com/OLEGSHA/PiwcsPrwModel/pull/1/commits/bc046b05dbee9d53399a493692ace0d66b15dfa6.

OLEGSHA commented on src/correctness.cpp:
```diff
@@ -0,0 +1,120 @@
+	isOutward.fill(false);
+
+	auto sectionAt = [&](SlotId slot) -> const Section*{
+	    return model.section(node->section(slot));
```
Add a runtime assertion to ensure `model.section` does not return a `nullptr`

Lipovr replied:
I think it would be better to check if `model.section` returns `nullptr` and return `false` as a node with empty slot signifies that `model` is incomplete. See 574bd5a39fc3273895762c5b3d6b2128937c9202.

OLEGSHA replied:
That would be a good solution if we wanted to use exceptions. I'm kinda trying to avoid them. Sure, this one is internal, but I don't like the added complexity.

Perhaps it'd be easier to cache all sections at the beginning and return `false` there and then, since no lambdas would be involved.

Lipovr replied:
Hello, sorry for the long delay in working on this PR.
I have reverted 574bd5a39fc3273895762c5b3d6b2128937c9202 and implemented necessary changes in d3d4ade1d649bb73d87c8ea077680c4789cc50ac and 7cb61817f22cb23640dc7e049500b287ec24319f correspondingly. Now all section pointers are cached in `std::array` and on encountering `nullptr` among them function returns `false`. I also added runtime assertion to check if section id is `ID_INVALID` which should help to trace errors in case by-slot loop is incorrect.

OLEGSHA commented on test/correctness.cpp:
```diff
@@ -0,0 +1,422 @@
+        const SectionTestInfo (& sections)[N]
+        ){
+
+    assert(
```
Use `_ASSERT` from `src/debug.h`, clang-tidy won't be happy otherwise

Lipovr replied:
Adressed in a1b3961.

OLEGSHA commented on test/correctness.cpp:
```diff
@@ -0,0 +1,422 @@
+        ASSERT_EQ(res, true);                                       \
+    }                                                               \
+
+// << "Expected correct node of type `" << #nodeType << "`";
```
?

Lipovr replied:
Adressed in 620d601.

OLEGSHA commented on test/correctness.cpp:
```diff
@@ -0,0 +1,422 @@
+        sections[slot] = {SType::NONE};
+    }
+
+    for(NodeType type : types){
```
Inline `types` here

Lipovr replied:
My reasoning behind `types` being a variable was that we could need some sort of list of every possible `NodeType`. elsewhere in tests. It turns out such list is already inlined in [`io_read.cpp:48`](https://github.com/OLEGSHA/PiwcsPrwModel/blob/ab09de4b3232a70d08bee4840935827ab4a69a50/src/io_read.cpp#L48C29-L56C11). What are your thoughts on extracting it to a `const` variable in `nodes.h` where `NodeType` is declared?

Lipovr replied:
I inlined it (b379baf875cf048a901616f6d818882849a23027) for the time being.

OLEGSHA commented on test/correctness.cpp:
```diff
@@ -0,0 +1,422 @@
+        ASSERT_TRUE(res) << "Expected correct node of type `" << type->name << "`";
+    }
+}
+
```
Test for correct error handling in `isLocallyCorrect`: non-existing ID, `ID_NULL`, empty Model, incomplete Model

Lipovr replied:
Added tests for invalid ids and empty Model in 37e1a64336c8d284350b89fce390041d773c2a00. Test for incomplete Model (a node with empty slot) was added in 574bd5a39fc3273895762c5b3d6b2128937c9202 ([code](https://github.com/Lipovr/PiwcsPrwModel/blame/37e1a64336c8d284350b89fce390041d773c2a00/test/correctness.cpp#L67-L71C2)).

Lipovr committed:
Reformat with clang-format

Lipovr committed:
Changes proposed from https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1596401212

Lipovr committed:
Changes proposed from https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1596391699

Lipovr committed:
Changes proposed from https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1596410047

Lipovr committed:
Changes proposed from https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1596411948

Lipovr committed:
Revert changes rejected by maintainer in https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1597442948

Revert commit 574bd5a39fc3273895762c5b3d6b2128937c9202 except changes to tests.

Lipovr committed:
Make `isLocallyCorrect` return `false` on a node with empty slot

Implement changes proposed from https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1597442948
Cache section pointers
Check for invalid `SlotId`
Return `false` on section with invalid id or empty slot

Lipovr committed:
Inlined `types` as proposed from https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1596412934

Lipovr committed:
Add check for `nullptr` as discussed in https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1597441541

Lipovr committed:
Add tests for `isLocallyCorrect` to check for non-existing id, `ID_NULL`

Lipovr committed:
Changes proposed from https://github.com/OLEGSHA/PiwcsPrwModel/pull/1#discussion_r1596388514

- Use `isUnidirectional` flag to improve readability
- Change comments
- Remove unnecessary check duplicate

Lipovr committed:
Format recent changes with clang-format

Lipovr committed:
Move declaration of `isLocallyCorrect` to `algorithms.h`

- Move declaration of `isLocallyCorrect` to public header `algorithms.h`
- Add documentation for `isLocallyCorrect`
- Move section explaining local correctness to `isLocallyCorrect`

Lipovr committed:
Improve readability of `tests/correctness.cpp`

- Reformat to improve readability
- Add clang-format guards
